ENH make CalibratedClassifierCV accept on fit_params

BenjaminBossan opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->

Re-creating PR #15218, the reason being that I probably messed up the rebase
there, leading to 865 files being changed. Here is a clean PR.

Partly addresses #12384

#### What does this implement/fix? Explain your changes.

This PR makes it possible to pass `fit_params` to the fit method of the
`CalibratedClassifierCV`, which are then routed to the underlying base
estimator.

*Note*: I implemented `predict_proba` on `CheckingClassifier` for the
new unit test to work.

Commit:
CalibratedClassifierCV passes on fit_params

Partly addresses to https://github.com/scikit-learn/scikit-learn/issues/12325

This PR makes it possible to pass fit_params to the fit method of the
CalibratedClassifierCV, which are then routed to the underlying base
estimator.

Commit:
CalibratedClassifierCV passes on fit_params

Partly addresses to https://github.com/scikit-learn/scikit-learn/issues/12325

This PR makes it possible to pass fit_params to the fit method of the
CalibratedClassifierCV, which are then routed to the underlying base
estimator.

Commit:
Add test for CheckingClassifier

After adding the predict_proba method on CheckingClassifier, the test
coverage decreased. Therefore, a test for predict_proba was
added. More tests for CheckingClassifier will be added in a separate
PR.

Commit:
Add test for CheckingClassifier

After adding the predict_proba method on CheckingClassifier, the test
coverage decreased. Therefore, a test for predict_proba was
added. More tests for CheckingClassifier will be added in a separate
PR.

Commit:
Remove unnecessary test code

Commit:
Remove unnecessary test code

Commit:
Rename function used in test

Commit:
Rename function used in test

BenjaminBossan committed:
Address reviewer comments

* remove unnecessary check for empty dict
* rename variable name to canonical X from T

Commit:
Address reviewer comments

Use _safe_indexing to index into fit_params in case they are not numpy
arrays.

Commit:
Merge branch 'calibratedclassifiercv-passes-on-fit-params' of https://github.com/BenjaminBossan/scikit-learn into calibratedclassifiercv-passes-on-fit-params

Commit:
Merge branch 'calibratedclassifiercv-passes-on-fit-params' of https://github.com/BenjaminBossan/scikit-learn into calibratedclassifiercv-passes-on-fit-params

Commit:
Fix linting error, line too long

BenjaminBossan commented:
@cmarmo Thank you for reminding me, I forgot about the PR.

As mentioned above, I recreated the PR because of some rebasing issues. The [original PR](https://github.com/scikit-learn/scikit-learn/pull/15218) already had approval by ogrisel.

Apart from addressing the last reviewer comment by jnothman, the main work here was to integrate my changes with a recent PR that introduced parallel fitting in `CalibratedClassifierCV` via joblib, which necessitated a considerable rewrite of my changes.

BenjaminBossan commented:
An unrelated proposal from my side: We could add a `return self` after line 244:

https://github.com/scikit-learn/scikit-learn/blob/41d648ef8863e6cadcb6c0a13f66dda7d0e9ea89/sklearn/calibration.py#L244-L246

Using this early return, we can remove the `else` from line 245 and save one level of indentation spanning 50 LOC starting from line 246. If this is in line with the sklearn coding style, I could add that change on top of this PR.

thomasjpfan commented:
> Using this early return, we can remove the else from line 245 and save one level of indentation spanning 50 LOC starting from line 246. If this is in line with the sklearn coding style, I could add that change on top of this PR.

I think doing this style change in another PR would be better, so we do not increase the diff of this PR. On the proposal, I tend to prefer doing early returns for the same reason you stated.

Commit:
Update whats_new

BenjaminBossan commented:
@thomasjpfan I updated the `whats_new`. While doing that, I also removed the `sklearn.calibrator` section, which probably was intended to be `sklearn.calibration`

thomasjpfan commented on sklearn/calibration.py:
```diff
@@ -266,30 +281,32 @@ def fit(self, X, y, sample_weight=None):
                                   "sample weights will only be used for the "
                                   "calibration itself." % estimator_name)
+                else:
+                    sample_weight_base_estimator = sample_weight
```
In this specific case,  is `sample_weight_base_estimator` always equal to `sample_weight`? If this is the case, I do not think we need a `sample_weight_base_estimator` parameter in `_fit_calibrated_classifer`.

thomasjpfan commented on sklearn/calibration.py:
```diff
@@ -33,25 +33,36 @@
     """
-    if sample_weight is not None and supports_sw:
+    fit_params_train = {}
+    for key, val in fit_params.items():
+        check_consistent_length(y, val)
```
Can we do the `check_consistent_length` all at once in `fit`? This way the parallel calls do not need to validate anymore.

Commit:
Address reviewer comments:

* don't use sample_weight_base_estimator, since it's the same as
  sample_weight
* perform check_consistent_length only once, inside fit

In addition:

* added a test for check_consistent_length

BenjaminBossan committed:
Merge branch 'master' into calibratedclassifiercv-passes-on-fit-params

BenjaminBossan commented:
@thomasjpfan Very good comments, I addressed the issues you raised.

> In this specific case, is `sample_weight_base_estimator` always equal to `sample_weight`?

Indeed. I guess it would be possible to have different values if we allowed

```python
calibrated_classifier_cv.fit(X, y, sample_weight=sample_weight, base_estimator__sample_weight=other_sample_weight)
```

But that is probably overkill.

thomasjpfan reviewed (approved):
Otherwise LGTM

Commit:
Address reviewer comment

* Clearer way to define base_estimator_uses_sw

Commit:
Add a test that sample_weight is passed to the base_estimator

Make sure that sample_weight is correctly passed to the
base_estimator. For this CheckingClassifier had to be extended to
include a check for sample_weight.

Commit:
Merge branch 'calibratedclassifiercv-passes-on-fit-params' of https://github.com/BenjaminBossan/scikit-learn into calibratedclassifiercv-passes-on-fit-params

BenjaminBossan commented on sklearn/calibration.py:
```diff
@@ -266,30 +279,32 @@ def fit(self, X, y, sample_weight=None):
                                   "sample weights will only be used for the "
                                   "calibration itself." % estimator_name)
+                else:
+                    base_estimator_uses_sw = True
```
:+1:

BenjaminBossan commented:
@thomasjpfan I addressed your comment.

On top of that, I added two more tests, to check if `sample_weight` is correctly routed to the base estimator. This in turn required me to extend `CheckingClassifier` to add an option to check for `sample_weight`.

Commit:
Fix linter complaint

Commit:
Fix failing unit tests

Tests failed because of changes to CheckingClassifier now having a
separate expected_sample_weight argument, instead of using
expected_fit_params for this purpose.

thomasjpfan reviewed (approved):
Still LGTM

glemaitre commented on sklearn/calibration.py:
```diff
@@ -213,13 +219,20 @@ def fit(self, X, y, sample_weight=None):
         sample_weight : array-like of shape (n_samples,), default=None
             Sample weights. If None, then samples are equally weighted.
 
+        **fit_params : dict of string -> object
```
`dict of str`

BenjaminBossan replied:
:+1:

glemaitre commented on sklearn/calibration.py:
```diff
@@ -213,13 +219,20 @@ def fit(self, X, y, sample_weight=None):
         """
         X, y = indexable(X, y)
 
+        for key, val in fit_params.items():
```
We can delay this part until and use `sklearn.utils.validation._check_fit_params` and select index at this moment.

BenjaminBossan replied:
Not sure what you mean here.

glemaitre commented on sklearn/calibration.py:
```diff
@@ -266,12 +279,13 @@ def fit(self, X, y, sample_weight=None):
                 sample_weight = _check_sample_weight(sample_weight, X)
 
-                if not supports_sw:
+                if not base_estimator_supports_sw:
```
This is a bit weird here because we are rechecking for th outer condition

```python
if not base_estimator_supports_sw:
```

is enough

BenjaminBossan replied:
Do I understand correctly that you would like to move up the `if not base_estimator_supports_sw` check outside of `if sample_weight is not None`? If so, we would warn about not using `sample_weight` in the base estimator even if `sample_weight is None`, which I think would be unnecessary.

glemaitre commented on sklearn/calibration.py:
```diff
@@ -280,16 +294,16 @@ def fit(self, X, y, sample_weight=None):
-                                                    for train, test
-                                                    in cv.split(X, y))
-
+                _fit_calibrated_classifer)(
```
Let's correct the spelling:

`_fit_calibrated_classifier`

BenjaminBossan replied:
:+1:

glemaitre commented on sklearn/calibration.py:
```diff
@@ -33,25 +33,31 @@
 
-def _fit_calibrated_classifer(estimator, X, y, train, test, supports_sw,
-                              method, classes, sample_weight=None):
+def _fit_calibrated_classifer(estimator, X, y, train, test,
```
Let's correct he spelling:

```python
_fit_calibrated_classifier
```

BenjaminBossan replied:
:+1:

glemaitre commented on sklearn/calibration.py:
```diff
@@ -33,25 +33,31 @@
-    if sample_weight is not None and supports_sw:
-        estimator.fit(X[train], y[train],
-                      sample_weight=sample_weight[train])
+    fit_params_train = {}
```
Here, we can use `_check_fit_params`:

```python
fit_params_train = _check_fit_params(X, fit_params, train)
```

BenjaminBossan replied:
:+1:

glemaitre commented on sklearn/calibration.py:
```diff
@@ -33,25 +33,31 @@
-                              method, classes, sample_weight=None):
+def _fit_calibrated_classifer(estimator, X, y, train, test,
+                              method, classes, sample_weight=None,
+                              base_estimator_uses_sw=False, **fit_params):
```
I would avoid to pass `base_estimator_uses_sw`. I think that we can make the check within the function itself again.

BenjaminBossan replied:
:+1:

glemaitre commented on sklearn/calibration.py:
```diff
@@ -33,25 +33,31 @@
+    for key, val in fit_params.items():
+        fit_params_train[key] = _safe_indexing(val, train)
+
+    if base_estimator_uses_sw:
```
```python
X_train, y_train = _safe_indexing(X, train), _safe_indexing(y, train)
X_test, y_test = _safe_indexing(X, test), _safe_indexing(y, test)
if sample_weight is not None:
    sample_weight_train = _safe_indexing(sample_weight, train)
    sample_weight_test = _safe_indexing(sample_weight, test)
    fit_parameters = signature(base_estimator.fit).parameters
    if "sample_weight" in fit_parameters:
        estimator.fit(
            X_train, y_train, sample_weight=sample_weight_train,
            **fit_params_train,
        )
    else:
        estimator.fit(X_train, y_train, **fit_params_train)
calibrated_classifier = _CalibratedClassifier(
    estimator, method=method, classes=classes,
)
calibrated_classifier.fit(
    X_test, y_test, sample_weight=sample_weight_test
)
```

BenjaminBossan replied:
The snippet doesn't quite work:

1. if `sample_weight` _is_ None, `estimator` is never fitted
2. on the last line, you use `sample_weight_test`, which is undefined if `sample_weight` is None

I tried to refactor according to the "spirit" of the snippet.

glemaitre commented on sklearn/calibration.py:
```diff
@@ -33,25 +33,31 @@
                                                   classes=classes)
-    sw = None if sample_weight is None else sample_weight[test]
+    sw = None if sample_weight is None else _safe_indexing(sample_weight, test)
     calibrated_classifier.fit(X[test], y[test], sample_weight=sw)
```
we could actually split the data above in the same if/else

glemaitre replied:
I put it in the snippet

BenjaminBossan replied:
:+1:

glemaitre commented on sklearn/calibration.py:
```diff
@@ -280,16 +294,16 @@ def fit(self, X, y, sample_weight=None):
+                    method=self.method,
+                    classes=self.classes_,
+                    sample_weight=sample_weight,
+                    base_estimator_uses_sw=base_estimator_uses_sw,
```
I would avoid to pass this parameter

BenjaminBossan replied:
:+1:

glemaitre commented on sklearn/tests/test_calibration.py:
```diff
@@ -455,3 +456,89 @@ def test_calibration_attributes(clf, cv):
+
+@pytest.mark.parametrize('fit_params_as_list', [False, True])
+def test_calibration_with_fit_params(fit_params_as_list):
+    """Tests that fit_params are passed to the underlying base estimator
```
missing final full stop

BenjaminBossan replied:
:+1:

glemaitre commented on sklearn/tests/test_calibration.py:
```diff
@@ -455,3 +456,89 @@ def test_calibration_attributes(clf, cv):
+    """Tests that fit_params are passed to the underlying base estimator
+
+    https://github.com/scikit-learn/scikit-learn/issues/12384
+
```
You should remove the blank line

BenjaminBossan replied:
:+1:

glemaitre commented on sklearn/tests/test_calibration.py:
```diff
@@ -455,3 +456,89 @@ def test_calibration_attributes(clf, cv):
+@pytest.mark.parametrize('fit_params_as_list', [False, True])
+def test_calibration_with_fit_params(fit_params_as_list):
+    """Tests that fit_params are passed to the underlying base estimator
+
```
You can add:

```
Related issue:
```

BenjaminBossan replied:
:+1:

glemaitre commented on sklearn/tests/test_calibration.py:
```diff
@@ -455,3 +456,89 @@ def test_calibration_attributes(clf, cv):
+    X, y = make_classification(n_samples=n_samples, n_features=6,
+                               random_state=42)
+
+    if fit_params_as_list:
```
You can use `sklearn.utils._testing._convert_container` to avoid the if else:

```python
@pytest.mark.parametrize("fit_params_type", ["list", "array"])
...
...
fit_params = {
    "a": _convert_container(y), "y": _convert_container(y)
}
```

BenjaminBossan replied:
:+1:

glemaitre commented on sklearn/tests/test_calibration.py:
```diff
@@ -455,3 +456,89 @@ def test_calibration_attributes(clf, cv):
+def test_calibration_with_sample_weight_base_estimator(sample_weight):
+    """Tests that sample_weight is passed to the underlying base
+    estimator
+
```
remove the blank line

BenjaminBossan replied:
:+1:

glemaitre commented on sklearn/tests/test_calibration.py:
```diff
@@ -455,3 +456,89 @@ def test_calibration_attributes(clf, cv):
+])
+def test_calibration_with_sample_weight_base_estimator(sample_weight):
+    """Tests that sample_weight is passed to the underlying base
+    estimator
```
missing full stop

BenjaminBossan replied:
:+1:

glemaitre commented on sklearn/tests/test_calibration.py:
```diff
@@ -455,3 +456,89 @@ def test_calibration_attributes(clf, cv):
+
+    There should be a warning, since the sample_weight is not passed
+    on to the base_estimator.
+
```
Remove this blank line

BenjaminBossan replied:
:+1:

glemaitre commented on sklearn/tests/test_calibration.py:
```diff
@@ -455,3 +456,89 @@ def test_calibration_attributes(clf, cv):
+
+def test_calibration_without_sample_weight_base_estimator():
+    """Check that even if the base_estimator doesn't support
+    sample_weight, fitting with sample_weight still works
```
add full stop

BenjaminBossan replied:
:+1:

glemaitre commented on sklearn/tests/test_calibration.py:
```diff
@@ -455,3 +456,89 @@ def test_calibration_attributes(clf, cv):
+def test_calibration_with_fit_params_inconsistent_length():
+    """fit_params having different length than data should raise the
+    correct error message
+
```
remove the blank line

BenjaminBossan replied:
:+1:

glemaitre commented on sklearn/tests/test_calibration.py:
```diff
@@ -455,3 +456,89 @@ def test_calibration_attributes(clf, cv):
+
+def test_calibration_with_fit_params_inconsistent_length():
+    """fit_params having different length than data should raise the
+    correct error message
```
add full stop

BenjaminBossan replied:
:+1:

glemaitre commented on sklearn/tests/test_calibration.py:
```diff
@@ -455,3 +456,89 @@ def test_calibration_attributes(clf, cv):
+
+    msg = (
+        r"Found input variables with inconsistent numbers of "
+        r"samples: \[200, 5\]")
```
add a retunrn before `)`

BenjaminBossan replied:
:+1:

glemaitre commented on sklearn/utils/_mocking.py:
```diff
@@ -81,6 +81,9 @@ class CheckingClassifier(ClassifierMixin, BaseEstimator):
     expected_fit_params : list of str, default=None
         A list of the expected parameters given when calling `fit`.
 
+    expected_sample_weight : bool (default=False)
```
`bool, default=False`

BenjaminBossan replied:
:+1:

glemaitre commented on sklearn/utils/_mocking.py:
```diff
@@ -81,6 +81,9 @@ class CheckingClassifier(ClassifierMixin, BaseEstimator):
         A list of the expected parameters given when calling `fit`.
 
+    expected_sample_weight : bool (default=False)
+        Whether to check if a valid sample_weight was passed to `fit`.
```
add backticks around `sample_weight`

BenjaminBossan replied:
:+1:

glemaitre commented on sklearn/utils/_mocking.py:
```diff
@@ -196,6 +204,13 @@ def fit(self, X, y, **fit_params):
+        if self.expected_sample_weight:
+            if sample_weight is None:
+                raise AssertionError("Expected sample_weight to be passed")
+            if len(sample_weight) != _num_samples(X):
```
It would be best to use `sklearn.utils.validation._check_sample_weight`

BenjaminBossan replied:
:+1:

glemaitre reviewed (changes requested):
In general it looks good. I just added a couple of changes which can make the code a bit more readable.

Commit:
Merge branch 'master' into calibratedclassifiercv-passes-on-fit-params

Commit:
Address reviewer comments

* fix typo: _fit_calibrated_classifer -> _fit_calibrated_classifier
  (not my typo btw ;)
* refactor _fit_calibrated_classifier
* make use of _check_fit_params
* use _convert_container in test
* apply fixes to docstrings
* use _check_sample_weight in test

BenjaminBossan reviewed (commented):
Thanks @glemaitre I fixed the issues you found.

For some of your comments, I have some comments or questions myself, please take a look.

Commit:
Update error message in test_checking_classifier_fit_params

Commit:
Changes to CheckingClassifier required tests to be adopted

In particular, tests in test_validation.py needed to be changed.

Commit:
Merge branch 'master' into calibratedclassifiercv-passes-on-fit-params

Commit:
Remove spurious characters from change log

Commit:
(Re-)Add comma for readability

Commit:
Fix a bug when samle weights are None

Commit:
Fix failing unit tests

Commit:
Fix linting error

Too many blank lines

cmarmo commented:
Hi @BenjaminBossan do you mind moving your changelog from `v0.24.rst` to `v1.0.rst`? Thanks!

Commit:
Move changelog to v1.0.rst

BenjaminBossan commented:
> Hi @BenjaminBossan do you mind moving your changelog from `v0.24.rst` to `v1.0.rst`? Thanks!

Thanks for pointing that out @cmarmo, I made the change. Didn't know that 1.0 was coming :tada:

cmarmo commented:
@glemaitre do you mind checking if your comments have been addressed? Thanks!
@BenjaminBossan in the meanwhile some conflicts arose, do you mind fixing them? Thanks!

BenjaminBossan commented:
@cmarmo I can work on solving the conflicts but would like to wait for the review first, otherwise it might create more work than necessary.

sirhylle commented on sklearn/calibration.py:
```diff
@@ -328,6 +348,8 @@ def fit(self, X, y, sample_weight=None):
                     this_estimator.fit(X, y, sample_weight)
```
It seems we should avoid positional arguments here. If not, we get an error with SGDClassifier for example (third positional argument is coef_init and not sample_weight).
`this_estimator.fit(X = X, y = y, sample_weight = sample_weight)`

BenjaminBossan replied:
Yes, good catch, I fixed this line.

Commit:
Merge remote-tracking branch 'upstream/main' into calibratedclassifiercv-passes-on-fit-params

Also:

- Reviewer comment: pass sample_weight as keyword argument
- Remove unused imports

BenjaminBossan commented:
@cmarmo 

I merged the current main branch, resolving any existing conflicts.

The [failing test](https://github.com/scikit-learn/scikit-learn/pull/18170/checks?check_run_id=2420144081) seems to be unrelated to my change:

```
584     >>> from sklearn.linear_model import LinearRegression
585     >>> X = np.array([[1, 1], [1, 2], [2, 2], [2, 3]])
586     >>> # y = 1 * x_0 + 2 * x_1 + 3
587     >>> y = np.dot(X, np.array([1, 2])) + 3
588     >>> reg = LinearRegression().fit(X, y)
589     >>> reg.score(X, y)
590     1.0
591     >>> reg.coef_
592     array([1., 2.])
593     >>> reg.intercept_
Expected:
    3.0000...
Got:
    3.0
```

But perhaps I'm missing something.

cmarmo commented:
> The [failing test](https://github.com/scikit-learn/scikit-learn/pull/18170/checks?check_run_id=2420144081) seems to be unrelated to my change:

Indeed! See #19963.

adrinjalali commented:
Pinging @glemaitre if you still want this in v1.0 :)

BenjaminBossan commented:
Although it would be awesome to be on the contributors list for 1.0, there are too many merge conflicts and open questions for a quick resolution. There are probably more pressing changes than this :)

glemaitre commented:
I will review this today (and probably push something to solve at least the merge conflicts)

BenjaminBossan commented:
Thanks @glemaitre. I can also take a look at the merge conflicts, but that wouldn't be today anymore (more likely sometimes before end of week).

fingoldo commented:
When is this expected to be ready? I am currently having a problem when trying to pass eval_set to a CatboostClassifier wrapped into CalibratedClassifierCV. 
Without CalibratedClassifierCV, `pipe.fit(X_train,y_train,est__eval_set=(X_test_val,y_test_val),est__plot=True)` works fine.
With it, both est__eval_set and est__base_estimator__eval_set throw "TypeError: fit() got an unexpected keyword argument." I believe this pr will solve my problem...

glemaitre committed:
black on future conflicted files

glemaitre committed:
Merge remote-tracking branch 'origin/main' into pr/BenjaminBossan/18170

glemaitre committed:
iter

glemaitre committed:
doc

glemaitre committed:
DOC update changelog

glemaitre committed:
DOC update for numpydoc validation

glemaitre committed:
iter

jjerphan commented on doc/whats_new/v1.1.rst:
```diff
@@ -75,6 +75,11 @@ Changelog
   add this information to the plot.
   :pr:`21038` by :user:`Guillaume Lemaitre <glemaitre>`.
 
+- |Enhancement| `fit` method of
+  :class:`calibration.CalibratedClassifierCV` now supports passing
+  `fit_params`, which are routed to the `base_estimator`.
+  :pr:`18170` by :user:`Benjamin Bossan <BenjaminBossan>`.
```
```suggestion
- |Enhancement| :meth:`calibration.CalibratedClassifierCV.fit` now supports passing
  `fit_params`, which are routed to the `base_estimator`.
  :pr:`18170` by :user:`Benjamin Bossan <BenjaminBossan>`.
```

jjerphan commented on sklearn/calibration.py:
```diff
@@ -264,6 +265,10 @@ def fit(self, X, y, sample_weight=None):
         sample_weight : array-like of shape (n_samples,), default=None
             Sample weights. If None, then samples are equally weighted.
 
+        **fit_params : dict
+            Parameters passed to the `fit` method of the underlying
+            classifier.
```
Tiny semantic nitpick:
```suggestion
        **fit_params : dict
            Parameters to pass to the `fit` method of the underlying
            classifier.
```

jjerphan commented on sklearn/calibration.py:
```diff
@@ -274,6 +279,10 @@ def fit(self, X, y, sample_weight=None):
+        for sample_aligned_params in fit_params.values():
+            check_consistent_length(y, sample_aligned_params)
+
+        self.calibrated_classifiers_ = []
```
Why is `self.calibrated_classifiers_` defined here?

glemaitre replied:
Maybe an error from solving the merge conflict. I will have a look.

jjerphan commented on sklearn/calibration.py:
```diff
@@ -496,28 +516,28 @@ def _fit_classifier_calibrator_pair(
     sample_weight : array-like, default=None
         Sample weights for `X`.
 
+    **fit_params : dict of str -> object
+        Parameters passed to the ``fit`` method of the estimator
```
To be consistent:
```suggestion
    **fit_params : dict
        Parameters to pass to the `fit` method of the underlying
        classifier.
```

jjerphan reviewed (commented):
A few suggestions; apart from those, the core of this PR LGTM.

glemaitre committed:
iter

glemaitre committed:
Apply suggestions from code review

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

glemaitre committed:
iter

glemaitre committed:
Merge branch 'calibratedclassifiercv-passes-on-fit-params' of github.com:BenjaminBossan/scikit-learn into pr/BenjaminBossan/18170

glemaitre reviewed (approved):
I think that I am fine with the current implementation. @jjerphan do you want to have new look.

jjerphan commented on sklearn/tests/test_calibration.py:
```diff
@@ -918,6 +928,82 @@ def test_calibrated_classifier_cv_double_sample_weights_equivalence(method, ense
 
+@pytest.mark.parametrize("fit_params_type", ["list", "array"])
+def test_calibration_with_fit_params(fit_params_type, data):
+    """Tests that fit_params are passed to the underlying base estimator.
+
+    Related issue:
+    https://github.com/scikit-learn/scikit-learn/issues/12384
+    """
```
```suggestion
    """Tests that fit_params are passed to the underlying base estimator.

    Non-regression test for:
    https://github.com/scikit-learn/scikit-learn/issues/12384
    """
```

jjerphan reviewed (approved):
LGTM. Thank you for initiating this work, @BenjaminBossan; thank you, @glemaitre, for pursuing it.

glemaitre committed:
Update sklearn/tests/test_calibration.py

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

kabartay commented:
> When is this expected to be ready? I am currently having a problem when trying to pass eval_set to a CatboostClassifier wrapped into CalibratedClassifierCV. Without CalibratedClassifierCV, `pipe.fit(X_train,y_train,est__eval_set=(X_test_val,y_test_val),est__plot=True)` works fine. With it, both est__eval_set and est__base_estimator__eval_set throw "TypeError: fit() got an unexpected keyword argument." I believe this pr will solve my problem... I can see a big work has been done already, please let's not lose all that efforts guys and get it fixed and approved )

I have similar issue, didn't work for me. Similarly, it worked without CalibratedClassifierCV but not with it.

kabartay commented:
@glemaitre @thomasjpfan @jjerphan @BenjaminBossan 
I need some help with proper usage of fit_params in `Pipeline` and `CalibratedClassifierCV` (my `sklearn` version is `v0.24.2`)

For example, a `Pipeline` with simple preprocessing and `CalibratedClassifierCV` with some ANN as a `base_estimator` 
```
model_preprocessing = ("preprocessing", ColumnTransformer([
        ('categorical', 'passthrough', categoricals),]), categoricals),  
        ('numerical', Pipeline([("scaler", RobustScaler()),("imputer", SimpleImputer())]), numericals)),
        ('drop', 'drop', drops),
    ], remainder='drop'))
calibrated_classifier = ("calibrated_classifier", CalibratedClassifierCV(
    base_estimator=model_class(**model_parameters), method='isotonic', cv=5))
pipeline = Pipeline([model_preprocessing, calibrated_classifier])
```

Some grid search with `GridSearchCV`
`model = GridSearchCV(estimator=pipeline,param_grid=param_grid,cv=5, scoring='accuracy',refit=False, verbose=2)`

Then I try to add `fit_params`, for example, `epochs`, `patience`m etc.
```
fit_params = {}
fit_params['calibrated_classifier__base_estimator__epochs'] = 100
fit_params['calibrated_classifier__base_estimator__patience'] = 10
```
And do fitting which errors:
`model.fit(x_train, y_train, **fit_params)`

with `TypeError: fit() got an unexpected keyword argument 'epochs'`:
  `File "/home/utilisateur/anaconda3/lib/python3.7/site-packages/sklearn/utils/validation.py", line 63, in inner_f      return f(*args, **kwargs)`
  `File "/home/utilisateur/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_search.py", line 880, in fit         self.best_estimator_.fit(X, y, **fit_params)`
  `File "/home/utilisateur/anaconda3/lib/python3.7/site-packages/sklearn/pipeline.py", line 346, in fit         self._final_estimator.fit(Xt, y, **fit_params_last_step)`

cannot understand why it doesn't properly pass `fit_params` to underlying estimator
without `CalibratedClassifierCV` I run properly, for example if use this kinda pipeline
```
model_preprocessing = ("preprocessing", ColumnTransformer([
        ('categorical', 'passthrough', categoricals),]), categoricals),
        ('numerical', Pipeline([("scaler", RobustScaler()),("imputer", SimpleImputer())]), numericals)),
        ('drop', 'drop', drops),
    ], remainder='drop'))
basic_classifier = ("basic_classifier", model_class(**model_parameters))
pipeline = Pipeline([model_preprocessing,basic_classifier])
```
Might I use not proper sklearn version or passing wrong name (not sure issue is here, I tried various stuff)?
Thanks for help.

glemaitre commented:
`fit_params` will only be supported in scikit-learn 1.1 that is not released yet and will be release in the coming month.

kabartay commented:
> `fit_params` will only be supported in scikit-learn 1.1 has not been released yet and will be released in the coming month.

@glemaitre thanks for letting me know. Is there approximate time such as end of March, mid-April, end April?
Now trying to test v1.1.dev0 just cloning, building and trying to use.

kabartay commented:
@glemaitre what can be the reason that causes failing certain tests with TypeError
For example,
- test_linear_regression_positive
  - `sklearn/linear_model/tests/test_base.py:276:`
`TypeError: _nnls.nnls() missing required argument 'n' (pos 3)`
  - `sklearn/linear_model/_base.py:687: TypeError`

- test_linear_regression_positive_multiple_outcome
  - `sklearn/linear_model/tests/test_base.py:301:`
  `sklearn/linear_model/_base.py:690: in fit`
  `/home/.local/lib/python3.8/site-packages/joblib/parallel.py:1041: in __call__`
  `/home/.local/lib/python3.8/site-packages/joblib/parallel.py:859: in dispatch_one_batch`
  `/home/.local/lib/python3.8/site-packages/joblib/parallel.py:777: in _dispatch`
  `/home/.local/lib/python3.8/site-packages/joblib/_parallel_backends.py:572: in __init__`
  `/home/.local/lib/python3.8/site-packages/joblib/parallel.py:262: in __call__`
  `TypeError: _nnls.nnls() missing required argument 'n' (pos 3)` -> finally lead to this 
  - `sklearn/utils/fixes.py:118: TypeError`
  
- test_linear_regression_positive_vs_nonpositive
  - `sklearn/linear_model/tests/test_base.py:315:`
 `TypeError: _nnls.nnls() missing required argument 'n' (pos 3)`
  - `sklearn/linear_model/_base.py:687: TypeError`
 
- test_linear_regression_positive_vs_nonpositive_when_positive
  - `sklearn/linear_model/tests/test_base.py:331:`
 `TypeError: _nnls.nnls() missing required argument 'n' (pos 3)`
  - `sklearn/linear_model/_base.py:687: TypeError`
 
- test_estimatorclasses_positive_constraint
  - `sklearn/linear_model/tests/test_least_angle.py:609:`
  - `sklearn/linear_model/_least_angle.py:2237: in fit`
  - `sklearn/linear_model/_least_angle.py:2286: in _estimate_noise_variance`
 `TypeError: _nnls.nnls() missing required argument 'n' (pos 3)` -> finally lead to this
  - `sklearn/linear_model/_base.py:687: TypeError`
 
 **Totally 5 failed**
